cupy/indexing/iterate.py

```diff
@@ -0,0 +1,139 @@
+import numpy
+
+import cupy
+from cupy import core
+from cupy.core import internal
+
+
+class flatiter():
+    """Flat iterator object to iterate over arrays.
+
+    A flatiter iterator is returned by ``x.flat`` for any array ``x``. It
+    allows iterating over the array as if it were a 1-D array, either in a
+    for-loop or by calling its ``next`` method.
+
+    Iteration is done in row-major, C-style order (the last index varying the
+    fastest).
+
+    .. note::
+       Restricted support of basic slicing is currently supplied. Advanced
+       indexing is not supported yet.
+
+    .. seealso:: :func:`numpy.flatiter`
+
+    """
+
+    def __init__(self, a):
+        self._base = a
+
+    def __setitem__(self, ind, value):
+        if ind is Ellipsis:
+            self[:] = value
+            return
+
+        if isinstance(ind, tuple):
+            raise IndexError('unsupported iterator index')
+
+        if isinstance(ind, bool):
+            raise IndexError('unsupported iterator index')
+
+        if numpy.isscalar(ind):
+            ind = int(ind)
+            base = self._base
+            size = base.size
+            indices = []
+            for s in base.shape:
+                size = size // s
+                indices.append(ind // size)
+                ind %= size
+            base[tuple(indices)] = value
+            return
+
+        if isinstance(ind, slice):
+            base = self._base
+            s = internal.complete_slice(ind, base.size)
+            s_start = s.start
```

Comment:
nit: use `size = s.stop - s.start` instead of tracking both

Reply:
Fixed!